docs/spec_draft.md

```diff
@@ -604,6 +605,86 @@ Afterwards, within each `process_group`:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the  values of the `operand` tensor from each process and
+produces a `result` tensor.
+
+The operation splits the StableHLO grid into process groups as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups)`.
+
+Afterwards, within each `process_group`:
+  * `operands@receiver = [operand@sender for sender in process_group]` for all
+    `receiver` in `process_group`.
+  * ```
+    result@process[i0, i1, ..., iR-1] =
+        reduce(
+          inputs=[operands@process[j][i0, i1, ..., iR-1] for j in range(size(operands@process))],
+          init_values=[I], # such that `computation` and `I` form a monoid]
+          dimensions=[0],
+          body=computation
+        )
+    ```
+
+### Inputs
+
+| Name                    | Type                                         |
+|-------------------------|----------------------------------------------|
+| `operand`               | tensor of any supported type                 |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64` |
```

Comment:
As discussed offline, this can be reformulated as "variadic number of 1-dimensional tensor constants of type `si64`".

Reply:
opened : https://github.com/openxla/stablehlo/issues/685

Let us keep it like this.

Reply:
How about we use this wording in the spec, and the ODS can follow suit later via the ticket?

Reply:
updated with 
https://github.com/openxla/stablehlo/issues/289